src/Symfony/Component/Translation/Bridge/Crowdin/CrowdinProvider.php

```diff
@@ -0,0 +1,397 @@
+            throw new ProviderException(sprintf('No content in exported file %d.', $fileId), $response);
+        }
+
+        if (200 !== $response->getStatusCode()) {
+            throw new ProviderException(sprintf('Unable to export file %d translations for language "%s".', $fileId, $languageId), $response);
+        }
+
+        $export = json_decode($response->getContent(), true);
+
+        return file_get_contents($export['data']['url']);
+    }
+
+    private function downloadSourceFile(int $fileId): string
+    {
+        /**
+         * @see https://support.crowdin.com/api/v2/#operation/api.projects.files.download.get (Crowdin API)
+         * @see https://support.crowdin.com/enterprise/api/#operation/api.projects.files.download.get (Crowdin Enterprise API)
+         */
+        $response = $this->client->request('GET', sprintf('https://%s/projects/%d/files/%d/download', $this->endpoint, $this->projectId, $fileId), [
+            'headers' => [
+                'Content-Type' => 'application/json',
+            ],
+        ]);
+
+        if (200 !== $response->getStatusCode()) {
+            throw new ProviderException(sprintf('Unable to download source file %d.', $fileId), $response);
+        }
+
+        $export = json_decode($response->getContent(), true);
+
+        return file_get_contents($export['data']['url']);
+    }
+
+    private function listStrings(int $fileId, int $limit, int $offset): array
+    {
+        /**
+         * @see https://support.crowdin.com/api/v2/#operation/api.projects.strings.getMany (Crowdin API)
+         * @see https://support.crowdin.com/enterprise/api/#operation/api.projects.strings.getMany (Crowdin Enterprise API)
+         */
+        $response = $this->client->request('GET', sprintf('https://%s/projects/%d/strings', $this->endpoint, $this->projectId), [
+            'query' => [
+                'fileId' => $fileId,
+                'limit' => $limit,
+                'offset' => $offset,
+            ],
+        ]);
+
+        if (200 !== $response->getStatusCode()) {
+            throw new ProviderException(sprintf('Unable to list strings for file %d in project %d. Message: "%s".', $fileId, $this->projectId, $response->getContent()), $response);
+        }
+
+        return json_decode($response->getContent(), true)['data'];
+    }
+
+    private function deleteString(int $stringId): void
+    {
+        /**
+         * @see https://support.crowdin.com/api/v2/#operation/api.projects.strings.delete (Crowdin API)
+         * @see https://support.crowdin.com/enterprise/api/#operation/api.projects.strings.delete (Crowdin Enterprise API)
+         */
+        $response = $this->client->request('DELETE', sprintf('https://%s/projects/%d/strings/%d', $this->endpoint, $this->projectId, $stringId));
+
+        if (404 === $response->getStatusCode()) {
+            $this->logger->warning(sprintf('String ID %d Not Found for the project %d.', $stringId, $this->projectId));
```

Comment:
should not be a warning. It is not an issue if the string is already removed.

Reply:
> Is it ok if the logger will not be used?

if the logger is not used, don't inject it at all.

> In what cases will be better to use a logger? Maybe instead of throwing some exceptions?

Logging is meant to help debugging by giving more details. Here, not finding an id that you want to delete is not an issue at all (after all, the _result_ you want is that the id does not exist anymore). So a log message is not really necessary, and even less a _warning_ message.